keychain_win.cpp

```diff
@@ -0,0 +1,211 @@
+    std::string errMsg;
+    LPWSTR errBuffer = nullptr;
+    ::FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
+                    nullptr, // ignored for the flags we use
+                    errorCode,
+                    0, // figure out LANGID automatically
+                    reinterpret_cast<LPWSTR>(&errBuffer),
+                    0,        // figure out out-buffer size automatically
+                    nullptr); // no additional arguments
+    if (errBuffer != nullptr) {
+        errMsg = wideCharToAnsi(errBuffer);
+        LocalFree(errBuffer);
+    }
+    return errMsg;
+}
+
+std::string makeTargetName(const std::string &package,
+                           const std::string &service,
+                           const std::string &user) {
+    return package + "." + service + '/' + user;
+}
+
+void updateError(keychain::Error &err) {
+    const auto code = ::GetLastError();
+    if (code == ERROR_SUCCESS) {
+        err = keychain::Error{};
+        return;
+    }
+
+    err.message = getErrorMessage(code);
+    err.code = code;
+    err.error = err.code == ERROR_NOT_FOUND
+                    ? keychain::KeychainError::NotFound
+                    : keychain::KeychainError::GenericError;
+}
+
+} // namespace
+
+namespace keychain {
+
+void setPassword(const std::string &package, const std::string &service,
+                 const std::string &user, const std::string &password,
+                 Error &err) {
+    ::SetLastError(0); // clear thread-global error
+
+    ScopedLpwstr target_name(
+        utf8ToWideChar(makeTargetName(package, service, user)));
+    if (!target_name) {
+        updateError(err);
+        return;
+    }
+
+    ScopedLpwstr user_name(utf8ToWideChar(user));
+    if (!user_name) {
+        updateError(err);
+        return;
+    }
+
+    if (password.size() > CRED_MAX_CREDENTIAL_BLOB_SIZE ||
+        password.size() > DWORD_MAX) {
+        err.error = KeychainError::PasswordTooLong;
+        err.message = "Password too long.";
+        err.code = -1; // generic non-zero
+        return;
+    }
+
+    CREDENTIAL cred = {0};
+    cred.Type = cred_type;
+    cred.TargetName = target_name.get();
+    cred.UserName = user_name.get();
+    cred.CredentialBlobSize = static_cast<DWORD>(password.size());
+    cred.CredentialBlob = (LPBYTE)(password.data());
+    cred.Persist = CRED_PERSIST_ENTERPRISE;
+
+    ::CredWrite(&cred, 0);
+    updateError(err);
+}
+
+std::string getPassword(const std::string &package, const std::string &service,
+                        const std::string &user, Error &err) {
+    ::SetLastError(0); // clear thread-global error
+    std::string password;
+
+    ScopedLpwstr target_name(
+        utf8ToWideChar(makeTargetName(package, service, user)));
+    if (!target_name) {
+        updateError(err);
+        return password;
+    }
+
+    CREDENTIAL *cred;
+    bool result = ::CredRead(target_name.get(), cred_type, 0, &cred);
+    updateError(err);
```

Comment:
only `updateError` if `result` tells you so?

Reply:
see discussion in `deletePassword`